### DISCUSSION TITLE
Architecture POC

### DISCUSSION THREAD

**Discussion by PangMo5** [jellyfin/Swiftfin#958] [Questions] [answered]
I saw that #957 was trying to restructure the current ViewModel, so I opened a discussion.

I personally would like to see the structure changed to [TCA](https://github.com/pointfreeco/swift-composable-architecture).

I think TCA has a lot of advantages that make it a good fit for our project.
such as the large number of developers using it and the low learning curve for new contributors to see our code due to its already established structure, which makes it easier for them to detect a Code Structure and feel comfortable contributing.
Of course, a pure View ViewModel pattern might have less of a learning curve, but I think it's hard to drive specific Code Structures and patterns.
And then there are the ancillary benefits: Swift-friendly patterns, State management, Convenient concurrency handling, `Composition` such as `ifLet` `switch` `forEach` reducer builders and operators, navigation, DI, performance, testability and the benefits of a unidirectional architecture

Personally, I've been using TCA on personal and company projects and I'm very satisfied with it.

As a side note, We can also take [TCA's Path Reducer](https://github.com/pointfreeco/swift-composable-architecture/blob/fd49eb28f52d734b3e19a503cf701d79397d3889/Examples/CaseStudies/SwiftUICaseStudies/03-NavigationStack.swift#L64) pattern to replace stinsen.
And in today's timely [update](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/migratingto1.7) made it easier to apply the [Observation](https://developer.apple.com/documentation/observation) tool. 😄 

Of course, changing all our code to be TCA-based can be a very long and arduous task.
But once you have a solid structure in place, it will be better for the future of your project.

If you're interested, I'll do my best to help.

**Comment by LePips** [comment, marked as answer]
I knew that it would only be a matter of time before someone suggested TCA, especially since I feel like it's gained a lot of prominence lately.

I've only personally dabbled in TCA and had a couple talks about it among my local dev group. It is great but very much has a steep learning curve, especially for beginners. It's also a framework that eventually becomes "all or nothing" in a project, which I'm fine with but that transition would be very arduous and distracting. I would only be open to such a large transition if we previously have a few people putting in multiple hours per week in the project and have enough knowledge of the app's features and design. I just frankly don't have the time or motivation to implement or oversee such a transition right now. I do think TCA is great and does come with a lot of features, I've specially noted [swift-perception](https://github.com/pointfreeco/swift-perception) which we could do as the first step towards TCA (since I think that can be used separately).

Edit:
This work would live on a separate branch and I'd be fine if someone wanted to transition to TCA on that branch (on this repo) with all the PRs where we can follow along and still be active in that transition. That would probably be more manageable for me and that would allow gauging the work and time expectations to determine feasibility.

---

On some specifics, we should eventually move away from Stinsen, however I strongly prefer a coordinator pattern over the way SwiftUI/TCA does it. I've looked at many repos that offer a similar pattern and may switch to one of those. Stinsen just hasn't received a lot of development lately (even though I have some rights on it so I could do it) but I like the simplicity of other packages that I've seen. The biggest kicker is that it still uses `NavigationView` but we will need to move on from iOS 15 at some point soon.

The `Stateful` for view models in #957 is just a rudimentary state holder and I don't even anticipate all view models needing to conform to it. Obviously properly adhering to a `Stateful` pattern can't be statically enforced but that's fine.
